### PR TITLE
Add DALI support.

### PR DESCRIPTION
**Pull request by gongweibao**
1. Add DALI support.
2. Fix NHWC pool shape bug.
3. Change reader from pyreader to dataloader.

### CODE REVIEW

File: benchmark/collective/resnet/scripts/train_gpu.sh [hunk 1]
```diff
@@ -28,20 +28,25 @@ DATA_PATH="./ImageNet"
...
 USE_HIERARCHICAL_ALLREDUCE=False
-NUM_CARDS=1
+NUM_CARDS=8
 FP16=True #whether to use float16
+use_dali=False
+if [[ ${use_dali} == "True" ]]; then
+    export FLAGS_fraction_of_gpu_memory_to_use=0.8
+    export FLAGS_conv_workspace_size_limit=4000 #MB
+fi
+export LD_LIBRARY_PATH=/root/go/soft/cuda10-cudnn7.6.5.32/lib64:${LD_LIBRARY_PATH}
```
Comment (gavin1332) [inline benchmark/collective/resnet/scripts/train_gpu.sh]: you'd better remove this environment

File: benchmark/collective/resnet/scripts/train_gpu.sh [hunk 2]
```diff
@@ -28,20 +28,25 @@ DATA_PATH="./ImageNet"
...
 
 # dgc params
 USE_DGC=False # whether to use dgc
-ALL_CARDS=8
+ALL_CARDS=1
```
Comment (gavin1332) [inline benchmark/collective/resnet/scripts/train_gpu.sh]: why change this parameter?

Comment (gavin1332) [inline benchmark/collective/resnet/scripts/train_gpu.sh]: why do you change this parameter?

File: benchmark/collective/resnet/dali.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+def convert_data_layout(data_layout):
+    if data_layout=='NCHW':
+        return types.NCHW
+    elif data_layout=='NHWC':
+        return types.NHWC
+    else:
+        assert False, "not supported data_layout:{}".format(data_layout)
+
+class HybridTrainPipe(Pipeline):
```
Comment (gavin1332) [inline benchmark/collective/resnet/dali.py:35]: some description is required to help understanding these code, as it is hard to GUESS what this pipeline stands for.

File: benchmark/collective/resnet/dali.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        images = self.decode(jpegs)
+        images = self.res(images)
+        output = self.cmnp(images.gpu(), mirror=rng)
+        return [output, self.to_int64(labels.gpu())]
+
+    def __len__(self):
+        return self.epoch_size("Reader")
+
+
+class HybridValPipe(Pipeline):
```
Comment (gavin1332) [inline benchmark/collective/resnet/dali.py:105]: same as above

File: benchmark/collective/resnet/scripts/train_gpu.sh [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
-
-export FLAGS_sync_nccl_allreduce=1
+export CUDA_VISIBLE_DEVICES=4,5,6,7
```
Comment (gavin1332) [inline benchmark/collective/resnet/scripts/train_gpu.sh]: wired environment settings, why not 0,1,2,3

File: benchmark/collective/resnet/utils/utility.py [hunk 6]
```diff
@@ -90,3 +93,57 @@ def get_median(data):
...
+        name="feed_y_a", shape=[None, 1], dtype="int64", lod_level=0)
+
+    if is_train and args.use_mixup:
+        feed_y_b = fluid.data(
+            name="feed_y_b", shape=[None, 1], dtype="int64", lod_level=0)
+        feed_lam = fluid.data(
+            name="feed_lam", shape=[None, 1], dtype="float32", lod_level=0)
+
+        data_loader = fluid.io.DataLoader.from_generator(
+            feed_list=[feed_image, feed_y_a, feed_y_b, feed_lam],
```
Comment (gavin1332) [inline benchmark/collective/resnet/utils/utility.py:134]: what y_b and y_b stand for?

**Review by gavin1332 — COMMENTED** [review]
require feedback

File: benchmark/collective/resnet/scripts/train_gpu.sh (hunk 5 above)
Comment (gongweibao) [reply benchmark/collective/resnet/scripts/train_gpu.sh]: These test lines will cleanup. Thanks!

File: benchmark/collective/resnet/scripts/train_gpu.sh (hunk 2 above)
Comment (gongweibao) [reply benchmark/collective/resnet/scripts/train_gpu.sh]: These test lines will cleanup. Thanks!

File: benchmark/collective/resnet/dali.py (hunk 3 above)
Comment (gongweibao) [reply benchmark/collective/resnet/dali.py:35]: It's from the DALI document. Will add some comments.Thanks.

File: benchmark/collective/resnet/dali.py (hunk 4 above)
Comment (gongweibao) [reply benchmark/collective/resnet/dali.py:105]: It's from the DALI document. Will add some comments.Thanks.

File: benchmark/collective/resnet/dali.py (hunk 3 above)
Comment (gongweibao) [reply benchmark/collective/resnet/dali.py:35]: Done.

File: benchmark/collective/resnet/utils/utility.py (hunk 6 above)
Comment (gongweibao) [reply benchmark/collective/resnet/utils/utility.py:134]: y_b is the second image's label.

File: benchmark/collective/resnet/dali.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+                 interp,
+                 mean,
+                 std,
+                 device_id,
+                 shard_id=0,
+                 num_shards=1,
+                 random_shuffle=True,
+                 num_threads=4,
+                 seed=42,
+                 data_layout="NCHW"):
```
Comment (wangxicoding) [inline benchmark/collective/resnet/dali.py:59]: too many parameters， maybe classify （file_root、file_list...) into file、(resize、crop...) into preprocessor is better?

**Review by gavin1332 — APPROVED** [review]
LGTM

File: benchmark/collective/resnet/dali.py (hunk 7 above)
Comment (gongweibao) [reply benchmark/collective/resnet/dali.py:59]: Find tune later.

**Pull request merged**